### DISCUSSION TITLE
[Bug]: react-hook-form force you to fill a disabled/readOnly field

### DISCUSSION THREAD

**Discussion by kimgiutae** [react-hook-form/react-hook-form#6296] [Q&A] [answered]
### Version Number

7.12.2

### Codesandbox/Expo snack

https://codesandbox.io/s/react-hook-form-v7-form-context-ytudi

### Steps to reproduce

1. Go to https://codesandbox.io/s/react-hook-form-v7-form-context-ytudi
2. Update all deps to latest stable
3. Go to the file Test.js
4. Add required rule to the register function inside input
5. Add readOnly or disabled in that input
6. Try to submit
7. Can't submit... the disabled input is required (?)

### Expected behaviour

The form should submit even if a required field now is disabled.

### What browsers are you seeing the problem on?

Edge

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by bluebill1049** [comment]
the question is why is `disabled` combined with `required`?

**Reply by kimgiutae** [reply]
Well... sometimes the disabled/readOnly prop is dynamic. depending in some other logic.

In my actual project, all fields are inside 3 fieldsets.

To go to the 2nd fieldset the first must validate first.

So the second fieldset must be disabled until the first fieldset validate. Or the fields inside 2nd fieldset must be readonly until all fields inside 1st fieldset validate.

**Reply by kimgiutae** [reply]
Other examples can be: a required field that appears when a checkbox is checked.

**Reply by bluebill1049** [reply]
i see what you mean now. At the moment, disabled at the input attribute is static, once you set it to true, hook form wouldn't be able to detect the change.  The next release will include a `disabled` prop at the register level, it will solve the problem. 7.13.0 it's a big release and it will be mostly like today.

**Reply by bluebill1049** [reply]
> Other examples can be: a required field that appears when a checkbox is checked.

appeared is a dependency display toggle not disabled toggle

**Reply by kimgiutae** [reply]
> i see what you mean now. At the moment, disabled at the input attribute is static, once you set it to true, hook form wouldn't be able to detect the change. The next release will include a `disabled` prop at the register level, it will solve the problem. 7.13.0 it's a big release and it will be mostly like today.

Many thanks. Seems that this won't work with dynamic disabled fieldset

**Reply by bluebill1049** [reply]
I think it will, you just have to pass the prop to the input as well. 

`<fieldset disabled> <input {...register('test', {disabled: true})} />`

**Reply by bluebill1049** [reply]
in fact, try it out now: 7.13.0-next.5

**Reply by kimgiutae** [reply]
> in fact, try it out now: 7.13.0-next.5

I try it on the same sandbox above updating to v7.13.0-next.5.

Added the disabled: true, but it doesn't submit. Keeps waiting to fill that field.

It only submit if I remove the required rule, but in that case is useless

**Reply by kimgiutae** [reply]
Anyways I wonder how can I make that disabled rule dynamic.... I have to re-register the field with new options?

**Reply by bluebill1049** [reply]
did you add disabled to the register as param? regsiter('test', { disabled: ??})

**Reply by kimgiutae** [reply]
> did you add disabled to the register as param? regsiter('test', { disabled: ??})

Yes

`return <input {...methods.register("bill", {
  required: true,
  disabled: true
})} />;`

BTW, the native behaviour of web forms allows to submit even if a required field is disabled.

**Reply by kimgiutae** [reply]
I know that must be complex into the core functionallity.

But maybe a solution can be... when updating formState on submitting attempt... just ignore the fields that are now disabled/readOnly, and remove errors from formState/instance.

On other hand... **it doesn't seems logic to put disabled as a rule**. Rules are often static. But the whole point on disabling something is because you have to make something first to enabling. Why show it then?

_Many thanks for the attention to my problem_

**Reply by bluebill1049** [reply]
That's valid. I will get a PR to patch this. disabled input will abort validation.

**Comment by bluebill1049** [comment, marked as answer]
This issue will get patching in the next release, disabled input will skip the validation.

**Reply by kimgiutae** [reply]
Many thanks for building this amazing tool

**Comment by kimgiutae** [comment]
I should add the disabled prop on register... or this should be detected on input prop change?

If it is the first case... I have to unregister and register with disabled prop on/off ?

Also, I saw the logic in last commit, and seems to get the disabled state from register... in that case it won't detect when ref prop updates?

**Reply by bluebill1049** [reply]
at `reigster` api not as input attribute

**Reply by kimgiutae** [reply]
So. How can I tell to react-hook-form that certain field disabled prop has changed?

As far as I know, register only occurs once on mounting...

**Reply by bluebill1049** [reply]
`register` is a function, it will get called again once the component re-render. it's really cheap to call register anyhow.

**Reply by kimgiutae** [reply]
Excellent. I'll test the new version and report any issue. Many thanks
